File: src/lib/processingEstimate.ts
/** Presentation only. All workload and queue arithmetic belongs to the server. */
export interface ProcessingEstimate {
  state: "range" | "queue_only" | "unknown" | "overdue";
  observed_at: string;
  expires_at: string;
  ready_earliest_at: string | null;
  ready_latest_at: string | null;
  start_earliest_at: string | null;
  start_latest_at: string | null;
  basis: string | null;
  reason: string | null;
}

export function formatProcessingEstimate(value: unknown, context: {
  now?: number; jobStatus: string | null; serviceState: string | undefined;
}): { summary: string; detail: string } | null {
  if (context.serviceState !== "available" || !["queued", "processing"].includes(context.jobStatus ?? "")
      || !value || typeof value !== "object") return null;
  const row = value as Record<string, unknown>;
  const now = context.now ?? Date.now();
  const date = (v: unknown) => typeof v === "string" ? Date.parse(v) : NaN;
  const observed = date(row.observed_at), expires = date(row.expires_at);
  if (!Number.isFinite(observed) || !Number.isFinite(expires) || observed > now + 30_000
      || now - observed > 90_000 || expires <= now || expires <= observed) return null;
  const overdue = { summary: "Processing is taking longer than estimated.", detail: "Your video is still in the processing queue. You can check back later." };
  if (row.state === "overdue") return overdue;
  if (row.state !== "range" && row.state !== "queue_only") return null;
  const queuedOnly = row.state === "queue_only";
  const low = date(queuedOnly ? row.start_earliest_at : row.ready_earliest_at);
  const high = date(queuedOnly ? row.start_latest_at : row.ready_latest_at);
  if (!Number.isFinite(low) || !Number.isFinite(high) || low > high) return null;
  if (high <= now) return queuedOnly ? null : overdue;
  const lower = Math.max(0, Math.floor((low - now) / 300_000) * 5);
  const upper = Math.max(5, Math.ceil((high - now) / 300_000) * 5);
  const range = lower === 0 ? `within about ${upper} minutes`
    : lower === upper ? `in about ${upper} minutes` : `in about ${lower}–${upper} minutes`;
  return queuedOnly ? {
    summary: `Estimated wait before processing: ${range.replace(/^in /, "")}.`,
    detail: row.reason === "metadata_unknown" ? "Processing time will be estimated after the video check."
      : "Processing time is not available yet.",
  } : {
    summary: `Estimated ready ${range}.`,
    detail: "This is a rough estimate based on recent processing.",
  };
}

File: ios/PongLens/PongLens/Core/ProcessingEstimate.swift
import Foundation

struct ProcessingEstimate: Decodable, Hashable {
    let readyScope: String?
    let state: String?
    let observedAt: String?
    let expiresAt: String?
    let readyEarliestAt: String?
    let readyLatestAt: String?
    let startEarliestAt: String?
    let startLatestAt: String?
    let basis: String?
    let reason: String?
    enum CodingKeys: String, CodingKey {
        case readyScope = "ready_scope"
        case state, basis, reason
        case observedAt = "observed_at", expiresAt = "expires_at"
        case readyEarliestAt = "ready_earliest_at", readyLatestAt = "ready_latest_at"
        case startEarliestAt = "start_earliest_at", startLatestAt = "start_latest_at"
    }
    struct Message { let summary: String; let detail: String }
    func message(jobStatus: String?, serviceState: String?, now: Date = Date()) -> Message? {
        guard readyScope == "match", state != "queue_only" else { return nil }
        guard serviceState == "available", jobStatus == "queued" || jobStatus == "processing" else { return nil }
        let iso = ISO8601DateFormatter()
        func date(_ text: String?) -> Date? {
            guard let text else { return nil }
            iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            if let value = iso.date(from: text) { return value }
            iso.formatOptions = [.withInternetDateTime]
            return iso.date(from: text)
        }
        guard let observed = date(observedAt), let expires = date(expiresAt),
              observed.timeIntervalSince(now) <= 30, now.timeIntervalSince(observed) <= 90,
              expires > now, expires > observed else { return nil }
        let overdue = Message(summary: "Processing is taking longer than estimated.",
                              detail: "Your video is still in the processing queue. You can check back later.")
        if state == "overdue" { return overdue }
        guard state == "range" || state == "queue_only" else { return nil }
        let queueOnly = state == "queue_only"
        guard let low = date(queueOnly ? startEarliestAt : readyEarliestAt),
              let high = date(queueOnly ? startLatestAt : readyLatestAt), low <= high else { return nil }
        if high <= now { return queueOnly ? nil : overdue }
        let lower = max(0, Int(floor(low.timeIntervalSince(now) / 300)) * 5)
        let upper = max(5, Int(ceil(high.timeIntervalSince(now) / 300)) * 5)
        let range = lower == 0 ? "within about \(upper) minutes"
            : lower == upper ? "in about \(upper) minutes" : "in about \(lower)–\(upper) minutes"
        if queueOnly {
            let wait = range.hasPrefix("in ") ? String(range.dropFirst(3)) : range
            return Message(summary: "Estimated wait before processing: \(wait).",
                           detail: reason == "metadata_unknown" ? "Processing time will be estimated after the video check." : "Processing time is not available yet.")
        }
        return Message(summary: "Estimated ready \(range).", detail: "This is a rough estimate based on recent processing.")
    }
}
